src/DataStreams/PushingToViewsBlockOutputStream.cpp

```diff
@@ -169,39 +221,32 @@ void PushingToViewsBlockOutputStream::write(const Block & block)
             output->write(block);
     }
 
-    /// Don't process materialized views if this block is duplicate
-    if (!getContext()->getSettingsRef().deduplicate_blocks_in_dependent_materialized_views && replicated_output && replicated_output->lastBlockIsDuplicate())
+    if (views.empty())
         return;
 
-    // Insert data into materialized views only after successful insert into main table
+    /// Don't process materialized views if this block is duplicate
     const Settings & settings = getContext()->getSettingsRef();
-    if (settings.parallel_view_processing && views.size() > 1)
+    if (!settings.deduplicate_blocks_in_dependent_materialized_views && replicated_output && replicated_output->lastBlockIsDuplicate())
+        return;
+
+    const size_t max_threads = std::min(views.size(), (settings.parallel_view_processing ? static_cast<size_t>(settings.max_threads) : 1));
```

Comment:
Note, that settings.max_threads could be zero, which means "unlimited". So we can use as many threads as we want, but more reasonably to use views.size() amount of threads.

Reply:
Great point. I'll fix it, as with the current change it would be processed sequentially.

Reply:
>Note, that settings.max_threads could be zero

Could it? In case of `max_threads=0` it should fallback to `getNumberOfPhysicalCPUCores()`
https://github.com/ClickHouse/ClickHouse/blob/c4a14ffca214dd876cedab8d0ec80046cc5ad006/src/Core/SettingsFields.h#L72